sycl/include/sycl/esimd/detail/esimd_types.hpp

```diff
@@ -0,0 +1,249 @@
+
+template <typename Ty, int N, typename EltTy, int Height, int Width>
+struct compute_format_type_2d<simd<Ty, N>, EltTy, Height, Width> {
+  static constexpr int Prod = sizeof(Ty) * N / sizeof(EltTy);
+  static_assert(Prod == Width * Height, "size mismatch");
+
+  static constexpr int SizeX = Width;
+  static constexpr int StrideX = 1;
+  static constexpr int SizeY = Height;
+  static constexpr int StrideY = 1;
+  using type = region2d_t<EltTy, SizeY, StrideY, SizeX, StrideX>;
+};
+
+template <typename BaseTy, typename RegionTy, typename EltTy, int Height,
+          int Width>
+struct compute_format_type_2d<simd_view<BaseTy, RegionTy>, EltTy, Height,
+                              Width> {
+  using ShapeTy = typename shape_type<RegionTy>::type;
+  static constexpr int Prod = ShapeTy::Size_in_bytes / sizeof(EltTy);
+  static_assert(Prod == Width * Height, "size mismatch");
+
+  static constexpr int SizeX = Width;
+  static constexpr int StrideX = 1;
+  static constexpr int SizeY = Height;
+  static constexpr int StrideY = 1;
+  using type = region2d_t<EltTy, SizeY, StrideY, SizeX, StrideX>;
+};
+
+template <typename Ty, typename EltTy, int Height, int Width>
+using compute_format_type_2d_t =
+    typename compute_format_type_2d<Ty, EltTy, Height, Width>::type;
+
+// Check if a type is simd_view type
+template <typename Ty> struct is_simd_view_type : std::false_type {};
+
+template <typename BaseTy, typename RegionTy>
+struct is_simd_view_type<simd_view<BaseTy, RegionTy>> : std::true_type {};
+
+template <typename Ty>
+struct is_simd_view_v
+    : std::integral_constant<bool,
+                             is_simd_view_type<remove_cvref_t<Ty>>::value> {};
+
+// Check if a type is simd or simd_view type
+template <typename Ty> struct is_simd_type : std::false_type {};
+
+template <typename Ty, int N>
+struct is_simd_type<simd<Ty, N>> : std::true_type {};
+
+template <typename BaseTy, typename RegionTy>
+struct is_simd_type<simd_view<BaseTy, RegionTy>> : std::true_type {};
+
+template <typename Ty>
+struct is_simd_v
+    : std::integral_constant<bool, is_simd_type<remove_cvref_t<Ty>>::value> {};
+
+// Get the element type if it is a simd or simd_view type.
+template <typename Ty> struct element_type { using type = remove_cvref_t<Ty>; };
+template <typename Ty, int N> struct element_type<simd<Ty, N>> {
+  using type = Ty;
+};
+template <typename BaseTy, typename RegionTy>
+struct element_type<simd_view<BaseTy, RegionTy>> {
+  using type = typename RegionTy::element_type;
+};
+
+// Get the common type of a binary operator.
+template <typename T1, typename T2,
+          typename =
+              csd::enable_if_t<is_simd_v<T1>::value && is_simd_v<T2>::value>>
+struct common_type {
+  using Ty1 = typename element_type<T1>::type;
+  using Ty2 = typename element_type<T2>::type;
+  using EltTy = decltype(Ty1() + Ty2());
+  static constexpr int N1 = T1::length;
+  static constexpr int N2 = T2::length;
+  static_assert(N1 == N2, "size mismatch");
```

Comment:
Are these values supposed to be private?

Reply:
Yes, they are. Fixed.

Reply:
Yes, they are. Fixed.